### PR TITLE
Dynamic prop support for UI Text Input

### PR DESCRIPTION
**Pull request by gayanSandamal**
## Description

This improvement includes dynamic property support for the `UI Text Input` widget, adding support for:

- Label
- Mode
- Icon
- IconPosition
- IconInnerPosition
- Clearable

## Related Issue(s)

https://github.com/FlowFuse/node-red-dashboard/issues/1177

## Checklist

 - [x] I have read the [contribution guidelines](https://github.com/FlowFuse/flowfuse/blob/main/CONTRIBUTING.md)
 - [ ] Suitable unit/system level tests have been added and they pass 
 - [ ] Documentation has been updated
    - [ ] Upgrade instructions
    - [ ] Configuration details
    - [ ] Concepts
 - [ ] Changes `flowforge.yml`?
    - [ ] Issue/PR raised on `FlowFuse/helm` to update ConfigMap Template
    - [ ] Issue/PR raised on `FlowFuse/CloudProject` to update values for Staging/Production

## Labels

 - [ ] Includes a DB migration? -> add the `area:migration` label

### CODE REVIEW

**Comment by gayanSandamal** [comment]
@joepavitt @Steve-Mcl 

Similar to the previous PR [#1173](https://github.com/FlowFuse/node-red-dashboard/pull/1173), I'm encountering a random test failure here. However, this time, it's consistently failing on test-input.spec.js on GitHub.

Strangely, when I run the tests locally using `npm run cy:run`, everything passes. I've run the tests multiple times to ensure this, and I've attached screenshots showing the successful runs.

_Note that the form.spec.js was failing on my local since the beginning as I discussed offline with @joepavitt _

Regarding all the above mentioned issues I'll keep this PR in draft mode and looking for any suggestions on how to proceed with this.

Screenshot 2024-08-18 at 21 51 41

Screenshot 2024-08-18 at 21 48 30

Screenshot 2024-08-18 at 21 44 50

**Comment by gayanSandamal** [comment]
@joepavitt I've marked this PR as "Ready to Review" because the written tests are working fine on my locall machine as the same as Number [Input Widget PR](https://github.com/FlowFuse/node-red-dashboard/pull/1173)

https://github.com/FlowFuse/node-red-dashboard/pull/1214#issuecomment-2295322627

File: cypress/tests/widgets/text-input.spec.js [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+        cy.get('#nrdb-ui-widget-8b7d7031372e4275').should('exist')
+        cy.clickAndWait(cy.get('button').contains('Dynamic Property: Label'))
+        cy.get('#nrdb-ui-widget-8b7d7031372e4275 .v-input .v-field .v-field__field').find('.v-field-label').should('exist')
+    })
+
+    it('Set the dynamic properties: set input "clearable"', () => {
+        cy.get('#nrdb-ui-widget-8b7d7031372e4275').should('exist')
+        cy.clickAndWait(cy.get('button').contains('Dynamic Property: Clearable'))
+        cy.get('#nrdb-ui-widget-8b7d7031372e4275').type('New text')
+        cy.get('#nrdb-ui-widget-8b7d7031372e4275 .v-input .v-field .v-field__field').find('.v-field-label').should('exist')
```
Comment (joepavitt) [inline cypress/tests/widgets/text-input.spec.js:45]: what are you actually testing here? It seems to me that it just checks that the `label` element exists, which isn't a useful check in content of whether the `clearable` function is available?

Comment (joepavitt) [reply cypress/tests/widgets/text-input.spec.js:45]: You should instead be checking for `v-field__clearable`

Comment (gayanSandamal) [reply cypress/tests/widgets/text-input.spec.js]: This is fixed but as usual tests are randomly failing on GH

Local test result 👇

Screenshot 2024-08-21 at 15 44 13

File: cypress/tests/widgets/text-input.spec.js [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+describe('Node-RED Dashboard 2.0 - Text Input (Dynamic Properties)', () => {
+    beforeEach(() => {
+        cy.deployFixture('dashboard-text-input')
+        cy.visit('/dashboard/page1')
+    })
+
+    it('Set the dynamic properties: set input "label"', () => {
+        cy.get('#nrdb-ui-widget-8b7d7031372e4275').should('exist')
+        cy.clickAndWait(cy.get('button').contains('Dynamic Property: Label'))
+        cy.get('#nrdb-ui-widget-8b7d7031372e4275 .v-input .v-field .v-field__field').find('.v-field-label').should('exist')
```
Comment (joepavitt) [inline cypress/tests/widgets/text-input.spec.js:38]: For this test too - what are you actually testing? The `label` will exist in all circumstances. The test is meant to check for a Dynamic label - so you should be checking your dynamic label has properly been set

Comment (gayanSandamal) [reply cypress/tests/widgets/text-input.spec.js:38]: Thanks for the feedback, Joe. Apologies for the oversight. I'll correct the test to ensure it checks the dynamic label and verifies that the text matches what's sent via injection. I'm working on it now.

**Comment by joepavitt** [comment]
> but the test we may need to abandon again, can't push this into production

@gayanSandamal okay to remove the dynamic properties E2E's please then we can merge this

**Comment by gayanSandamal** [comment]
> > but the test we may need to abandon again, can't push this into production
> 
> @gayanSandamal okay to remove the dynamic properties E2E's please then we can merge this

Thank you so much for doing the manual testing and verifying. I will remove E2E tests for dynamic properties and update this PR

**Comment by gayanSandamal** [comment]
> > but the test we may need to abandon again, can't push this into production
> 
> @gayanSandamal okay to remove the dynamic properties E2E's please then we can merge this

Removed the randomly failing dynamic prop E2E tests of the text-input widget and number-input as well because in there also the dynamic prop E2E tests are failing.

This is ready for merge

**Comment by joepavitt** [comment]
Like all of the other Dynamic Property PRs you've opened, there are no server-side changes, and so all dynamic properties are lost on refresh of the browser - please look at the other widgets already updated to see how this should be implemented

**Comment by gayanSandamal** [comment]
> Like all of the other Dynamic Property PRs you've opened, there are no server-side changes, and so all dynamic properties are lost on refresh of the browser - please look at the other widgets already updated to see how this should be implemented

@joepavitt applied the fix

**Comment by gayanSandamal** [comment]
> > Like all of the other Dynamic Property PRs you've opened, there are no server-side changes, and so all dynamic properties are lost on refresh of the browser - please look at the other widgets already updated to see how this should be implemented
> 
> @joepavitt applied the fix. Marking this as ready for review

File: docs/nodes/widgets/ui-text-input.md [hunk 3]
```diff
@@ -24,6 +36,24 @@ dynamic:
     Class:
         payload: msg.class
         structure: ["String"]
+    Mode:
+        payload: msg.mode
```
Comment (joepavitt) [inline docs/nodes/widgets/ui-text-input.md:40]:
```suggestion
        payload: msg.ui_update.mode
```

File: nodes/widgets/ui_text_input.js [hunk 4]
```diff
@@ -1,20 +1,48 @@
...
+                    if (typeof updates.iconPosition !== 'undefined') {
+                        // dynamically set "icon position" property
+                        statestore.set(group.getBase(), node, msg, 'iconPosition', updates.iconPosition)
+                    }
+                    if (typeof updates.iconInnerPosition !== 'undefined') {
+                        // dynamically set "icon inner position" property
+                        statestore.set(group.getBase(), node, msg, 'iconInnerPosition', updates.iconInnerPosition)
+                    }
+                }
+                msg = await appendTopic(RED, config, node, msg)
```
Comment (joepavitt) [inline nodes/widgets/ui_text_input.js:43]:
```suggestion
```

No need for this as it's done in the `ui-base` logic for `onChange` and `onAction` already, and we have no use for `onInput` (where the `appendTopic` doesn't run by default) in a Text Input node

**Review by joepavitt — CHANGES_REQUESTED** [review]

**Comment by joepavitt** [comment]
In [bf5c6f8](https://github.com/FlowFuse/node-red-dashboard/pull/1214/commits/bf5c6f8752dac72d24a23374dff109bbfdaaa15f) I have added:

- In Editor Documentation (was missing completely)
- Example for `iconPosition` and `iconInnerPosition` as no documentation to detail what values are expected there
- Correct in the in-Editor documentation to inform user properties need to be defined at `msg.ui_update.`, not `msg.`

**Pull request merged**